migrations/block_transactions.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/ledgerwatch/erigon/rlp"
+	"path/filepath"
+)
+
+var splitCanonicalAndNonCanonicalTransactionsBuckets = Migration{
+	Name: "split_canonical_and_noncanonical_txs",
+	Up: func(db ethdb.Database, tmpdir string, progress []byte, CommitProgress etl.LoadCommitHandler) (err error) {
+		bodiesPath := filepath.Join(tmpdir, "bodies")
+		ethtxPath := filepath.Join(tmpdir, "ethtx")
+		nonCanonicalPath := filepath.Join(tmpdir, "noncanonicaltx")
+		bodiesCollector := etl.NewCollector(bodiesPath, etl.NewSortableBuffer(etl.BufferOptimalSize))
+		ethTXCollector := etl.NewCollector(ethtxPath, etl.NewSortableBuffer(etl.BufferOptimalSize))
+		nonCanonicalCollector := etl.NewCollector(nonCanonicalPath, etl.NewSortableBuffer(etl.BufferOptimalSize))
+
+		ethTXWriteCursor, err := db.(ethdb.HasTx).Tx().(ethdb.RwTx).Cursor(dbutils.EthTx)
+		if err != nil {
+			return err
+		}
+		bfs := &types.BodyForStorage{}
+		var ethTXIndex, nonCanonicalIndex uint64
+
+		err = db.ForEach(dbutils.BlockBodyPrefix, []byte{}, func(k, v []byte) error {
+			blockNum := binary.BigEndian.Uint64(k[:8])
+			blockHash := common.BytesToHash(k[8:])
+			err = rlp.DecodeBytes(v, bfs)
+			if err != nil {
+				return err
+			}
+			canonicalHash, err := rawdb.ReadCanonicalHash(db, blockNum)
+			if err != nil {
+				return err
+			}
+			if blockHash == canonicalHash {
+				bfs.Canonical = true
+				if bfs.TxAmount > 0 {
+					txIdKey := make([]byte, 8)
+					binary.BigEndian.PutUint64(txIdKey, bfs.BaseTxId)
+					i := uint32(0)
+
+					for k, v, err := ethTXWriteCursor.SeekExact(txIdKey); k != nil; k, v, err = ethTXWriteCursor.Next() {
+						if err != nil {
+							return err
+						}
+
+						err = ethTXCollector.Collect(dbutils.EncodeBlockNumber(ethTXIndex+uint64(i)), common.CopyBytes(v))
+						if err != nil {
+							return err
+						}
+
+						i++
+						if i >= bfs.TxAmount {
+							break
+						}
+					}
+				}
+				bfs.BaseTxId = ethTXIndex
+				ethTXIndex += uint64(bfs.TxAmount)
+			} else {
+				bfs.Canonical = false
+				if bfs.TxAmount > 0 {
+					txIdKey := make([]byte, 8)
+					binary.BigEndian.PutUint64(txIdKey, bfs.BaseTxId)
+					i := uint32(0)
+
+					for k, v, err := ethTXWriteCursor.SeekExact(txIdKey); k != nil; k, v, err = ethTXWriteCursor.Next() {
+						if err != nil {
+							return err
+						}
+
+						err = nonCanonicalCollector.Collect(dbutils.EncodeBlockNumber(nonCanonicalIndex+uint64(i)), common.CopyBytes(v))
+						if err != nil {
+							return err
+						}
+
+						i++
+						if i >= bfs.TxAmount {
+							break
+						}
+					}
+				}
+				bfs.BaseTxId = nonCanonicalIndex
+				nonCanonicalIndex += uint64(bfs.TxAmount)
+			}
+			bodyBytes, err := rlp.EncodeToBytes(bfs)
+			if err != nil {
+				return err
+			}
+
+			err = bodiesCollector.Collect(common.CopyBytes(k), bodyBytes)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		err = db.(ethdb.BucketMigrator).ClearBucket(dbutils.EthTx)
```

Comment:
ideally need commit tx after clear bucket - otherwise free space can't be re-used by next load.
See example here: https://github.com/ledgerwatch/erigon/commit/11d1e9119e898989be305cb4dcd9b903416a2980

x

Reply:
fixed